### PR TITLE
feat: add push notifications

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

Implements push notifications for transaction events.

## How this PR fixes it

New "Notifications" settings sections have been added per-Safes/all added Safes, as well as a feature banner and associated tracking has been added. Subscribed categories dispatch push notifccations when the Safe UI is either out of focus or closed (but the browser is open).

The per-Safe "Notifications" section allows for notification subscription to all notification types, requiring a confirmatory signature as "Confirmation requests" are included. The global section is similar, subscribing to user selected (added) Safes with n number of signatures per selected chain.

Notifications are grouped into three categories: "Incoming transactions", "Outgoing transactions" and "Confirmation requests" according to [our Transaction Service webhooks](https://github.com/safe-global/safe-transaction-service/blob/master/safe_transaction_service/notifications/tasks.py).

- Incoming transactions
  - `INCOMING_ETHER`
  - `INCOMING_TOKEN`
- Outgoing transactions
  - `MODULE_TRANSACTION`
  - `EXECUTED_MULTISIG_TRANSACTION`
- Confirmation requests
  - `CONFIRMATION_REQUEST`

A banner is shown once per chain that has added Safes, subscribing to notifications for all added Safes on that chain.

User interaction within the UI is tracked directly. ~Due to technical limitations, direct notification interaction events (opening/closing) are cached and tracked when the UI is again opened.~ Moved to a separate PR https://github.com/safe-global/safe-wallet-web/pull/2500.

## How to test it

⚠️ Due to the complexity of this feature, I would advise explaining the feature on a call and maybe even testing it together. ⚠️

1. Registering for notifications should succeed, showing a notification and then displayind them accordingly when the Safe UI is not in focus or is closed (but the browser is open).
2. Unregistering a Safe should only unregister itself if there are others subscribed to on that chain or the entire device if it is the last Safe subscribed to on that chain.
3. Tracking of [user interaction](https://github.com/safe-global/safe-wallet-web/blob/ceaacff11584fbcebecfbd2f449c6a72b073825c/src/services/analytics/events/push-notifications.ts#L14-L95) should happen immediately.
4. ~Tacking of [notification iteraction](https://github.com/safe-global/safe-wallet-web/blob/ceaacff11584fbcebecfbd2f449c6a72b073825c/src/services/analytics/events/push-notifications.ts#L5-L13) should happen when the UI is reopened.~ (Moved to a separate PR https://github.com/safe-global/safe-wallet-web/pull/2500).
5. The MacOS info banner should only show on Mac devices.

If you want to reset your current registration, you need to:

- Reset browser notification permissions
- Unregister the `sw.js` service worker
- Delete all `firebase-*`/`notification-*` IndexedDB databases
image

If you want to show the subscription banner again, you need to delete the `SAFE_v2__dismissPushNotifications` `localStorage` key.

## Screenshots

image

image

image

image

Please note that these are slightly dated but just demonstrate the notification. See the [designs](https://www.figma.com/file/ma5syWag1CO8yVbuzqkK0u/Push-notifications?mode=dev).

notification

## Checklist
* [ ] I've tested the branch on mobile 📱
* [x] I've documented how it affects the analytics (if at all) 📊
* [x] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

File: src/components/settings/PushNotifications/logic.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+  if (!isGranted) {
+    alert('You must allow notifications to register your device.')
+  }
+
+  return isGranted
+}
+
+const getSafeRegistrationSignature = ({
+  safeAddresses,
```
Comment (schmanu) [inline src/components/settings/PushNotifications/logic.ts:42]: Is there a case where we have more than one Safe address for the registration that requires a signature?

File: src/hooks/useFirebaseNotifications.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      return
+    }
+
+    const hasFirebaseOptions = Object.values(FIREBASE_OPTIONS).every(Boolean)
+
+    if (!hasFirebaseOptions) {
+      return
+    }
+
+    initializeApp(FIREBASE_OPTIONS)
```
Comment (schmanu) [inline src/hooks/useFirebaseNotifications.ts:19]: Can this throw if some constants are configured wrong / change? Should we try / catch the initialization?

File: src/services/firebase/index.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+  const response = await fetch(ENDPOINT)
+
+  if (response.ok) {
+    return response.json()
+  }
+}
+
+const getLink = (path: string, query: { address: string; chain?: ChainInfo; safeTxHash?: string }) => {
+  const APP_URL = 'https://app.safe.global'
```
Comment (schmanu) [inline src/services/firebase/index.ts:65]: Can we somehow get the actual running app url? 
So localhost in dev / the PR deployments URLs?

File: src/services/firebase/index.ts [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    }
+  }
+
+  if (type === WebhookType.INCOMING_TOKEN || type === WebhookType.OUTGOING_TOKEN) {
+    const { tokenAddress, txHash, value } = data
+
+    let balances: SafeBalanceResponse | undefined
+
+    try {
+      balances = await getBalances(chainId, address)
```
Comment (schmanu) [inline src/services/firebase/index.ts:175]: Just a thought:
On tx service side there is an endpoint to fetch token infos:
i.e. https://safe-transaction-goerli.safe.global/api/v1/tokens/0x61fD3b6d656F39395e32f46E2050953376c3f5Ff

Should we add this to the gateway service at some point? There are multiple examples where we fetch balances only to get some token info.

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+          (safeAddress) => !selectedSafes[chainId]?.includes(safeAddress),
+        )
+
+        if (safesToUnregisterOnChain.length > 0) {
+          acc[chainId] = safeAddresses
+        }
+        return acc
+      }, {})
+
+    const shouldRegisterSafes = Object.values(safesToRegister).some((safeAddresses) => safeAddresses.length > 0)
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/GlobalPushNotifications.tsx:201]: The former is relevant to (un-)selected Safes. I will rename them accordingly.

File: src/components/settings/PushNotifications/logic.ts (hunk 1 above)
Comment (iamacook) [reply src/components/settings/PushNotifications/logic.ts:42]: No. In the current design, you can only sign for notifications within a Safe.

File: src/hooks/useFirebaseNotifications.ts (hunk 2 above)
Comment (iamacook) [reply src/hooks/useFirebaseNotifications.ts:19]: Good point. I'll wrap it in a `try / catch`.

File: src/services/firebase/index.ts (hunk 4 above)
Comment (iamacook) [reply src/services/firebase/index.ts:175]: This sounds like a great idea. Let's file a feature request for it on the gateway.

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+      return
+    }
+
+    const isGranted = await requestNotificationPermission()
+
+    if (!isGranted) {
+      return
+    }
+
+    // TODO: Enable when live on prod.
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/GlobalPushNotifications.tsx:167]: When I added this comment it wasn't released on the gateway. I'll investigate this now.

File: public/firebase-messaging-sw.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    false,
+  )
+
+  const app = initializeApp(FIREBASE_OPTIONS)
+
+  const messaging = getMessaging(app)
+
+  onBackgroundMessage(messaging, async (payload) => {
+    const ICON_PATH = '/images/safe-logo-green.png'
+    const DEFAULT_LINK = 'https://app.safe.global'
```
[the remark this answers is not in the document]
Comment (iamacook) [reply public/firebase-messaging-sw.ts]: `self.location.origin` is now used from 6f106769e7298409835617aa88d7d5a1d794f346.

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+      return Object.entries(notifiableSafes).reduce((acc, [chainId, safeAddresses]) => {
+        return {
+          ...acc,
+          [chainId]: safeAddresses,
+        }
+      }, {})
+    })
+  }
+
+  // Wether Safes need to be (un-)reigstered with the service
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/GlobalPushNotifications.tsx]: Fixed in 6f106769e7298409835617aa88d7d5a1d794f346.

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/GlobalPushNotifications.tsx]: It has been since deployed. Added functionality in 6f106769e7298409835617aa88d7d5a1d794f346.

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+
+    if (safesToUnregister) {
+      for (const [chainId, safeAddresses] of Object.entries(safesToUnregister)) {
+        for (const safeAddress of safeAddresses) {
+          registrationPromises.push(unregisterSafeNotifications(chainId, safeAddress))
+        }
+      }
+    }
+
+    Promise.all(registrationPromises)
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/GlobalPushNotifications.tsx]: Addressed in 6f106769e7298409835617aa88d7d5a1d794f346.

File: src/hooks/useFirebaseNotifications.ts (hunk 2 above)
Comment (iamacook) [reply src/hooks/useFirebaseNotifications.ts]: I abstracted the logic and used it in both places where we initialise Firebase in 6f106769e7298409835617aa88d7d5a1d794f346.

File: .github/workflows/build/action.yml [hunk 10]
```diff
@@ -43,3 +43,21 @@ runs:
...
+        NEXT_PUBLIC_FIREBASE_API_KEY_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_API_KEY_STAGING }}
+        NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN_STAGING }}
+        NEXT_PUBLIC_FIREBASE_DATABASE_URL_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_DATABASE_URL_STAGING }}
+        NEXT_PUBLIC_FIREBASE_PROJECT_ID_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_PROJECT_ID_STAGING }}
+        NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET_STAGING }}
+        NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID_STAGING }}
+        NEXT_PUBLIC_FIREBASE_APP_ID_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_APP_ID_STAGING }}
+        NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID_STAGING }}
+        NEXT_PUBLIC_FIREBASE_VAPID_KEY_STAGING: ${{ fromJSON(inputs.secrets).NEXT_PUBLIC_FIREBASE_VAPID_KEY_STAGING }}
```
Comment (katspaugh) [inline .github/workflows/build/action.yml:63]: That's a lot. Can we store them in a single env var as JSON?

File: next.config.mjs [hunk 11]
```diff
@@ -1,15 +1,6 @@
...
-import NextPwa from 'next-pwa'
-
-const withPWA = NextPwa({
-  disable: process.env.NODE_ENV === 'development',
-  dest: 'public',
-  reloadOnOnline: false,
-  /* Do not precache anything */
-  publicExcludes: ['**/*'],
-  buildExcludes: [/./],
-})
```
Comment (katspaugh) [inline next.config.mjs:12]: So there's no way to keep it?

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+
+  for (const { chainId, safeAddress } of Object.values(allPreferences)) {
+    if (!obj[chainId]) {
+      obj[chainId] = []
+    }
+
+    obj[chainId].push(safeAddress)
+  }
+
+  return obj
```
Comment (katspaugh) [inline src/components/settings/PushNotifications/GlobalPushNotifications.tsx:57]: In both of these a `reduce` would be a little easier to follow.

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+        for (const safeAddress of safeAddresses) {
+          registrationPromises.push(unregisterSafeNotifications(chainId, safeAddress))
+        }
+      }
+    }
+
+    await Promise.all(registrationPromises)
+
+    trackEvent(PUSH_NOTIFICATION_EVENTS.SAVE_SETTINGS)
+  }
```
Comment (katspaugh) [inline src/components/settings/PushNotifications/GlobalPushNotifications.tsx:229]: A lot is going on here and it's a bit hard to follow. Especially with regards to what functions need the state of the component and which are pure. I would suggest to extract as much logic as you can into pure functions outside of the component and keep JSX separate from logic.

**Review by katspaugh — COMMENTED** [review]
It looks great now in terms of what goes into the Service Worker. We could add a comment on top of each module used in the SW, saying something to the effect of "be careful what you import here".

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx (hunk 12 above)
Comment (iamacook) [reply src/components/settings/PushNotifications/GlobalPushNotifications.tsx]: Adjusted accordingly in 75b5a63d8189dbe37b42ffb469b1bb2361a2d28a.

File: src/components/settings/PushNotifications/GlobalPushNotifications.tsx (hunk 13 above)
Comment (iamacook) [reply src/components/settings/PushNotifications/GlobalPushNotifications.tsx]: I've abstracted the code outside of the component in 75b5a63d8189dbe37b42ffb469b1bb2361a2d28a.

**Comment by iamacook** [comment]
>We could add a comment on top of each module used in the SW, saying something to the effect of "be careful what you import here".

I've added comments in the relevant places.

> Also, you're probably aware, but there's now no tracking of clicks on notifications now (I think you had them in the previous version).

Now that I've addresses your review, I will now look our proposed solution of caching the events and dispatching them when the UI is opened.

File: src/tests/pages/apps-share.test.tsx [hunk 14]
```diff
@@ -4,16 +4,67 @@ import ShareSafeApp from '@/pages/share/safe-app'
 import { CONFIG_SERVICE_CHAINS } from '@/tests/mocks/chains'
 import * as useWalletHook from '@/hooks/wallets/useWallet'
 import * as useOwnedSafesHook from '@/hooks/useOwnedSafes'
+import * as manifest from '@/services/safe-apps/manifest'
+import * as sdk from '@safe-global/safe-gateway-typescript-sdk'
```
Comment (iamacook) [inline src/tests/pages/apps-share.test.tsx:8]: This test was breaking and I realised that none of the responses were mocked.

File: src/services/firebase/notifications.ts [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+
+  // Manually dispatched notifications from the Firebase admin panel
+  // Displayed as is
+  if (payload.notification) {
+    return {
+      title: payload.notification.title || '',
+      body: payload.notification.body,
+      image: payload.notification.image,
+    }
+  }
```
Comment (katspaugh) [inline src/services/firebase/notifications.ts:244]: I feel like this could be condensed into two-three lines?

File: src/components/settings/PushNotifications/logic.ts [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+  if (Notification.permission === 'granted') {
+    return true
+  }
+
+  let permission: NotificationPermission | undefined
+
+  try {
+    permission = await Notification.requestPermission()
+  } catch (e) {
+    console.error('Error requesting notification permission', e)
```
Comment (katspaugh) [inline src/components/settings/PushNotifications/logic.ts:27]: Any reason not to use the logError function here? Is this module running in the SW?

**Review by katspaugh — COMMENTED** [review]
I'm lacking clarity on which modules are used by a service worker and which are not. Can we either name them in a distinct way or maybe add a comment on top of each?

File: src/services/firebase/notifications.ts [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+
+  const DEFAULT_INFO = {
+    symbol: 'tokens',
+    value: 'some',
+    name: 'Token',
+  }
+
+  const tokenInfo = await getBalances(chainId, safeAddress, DEFAULT_CURRENCY)
+    .then(({ items }) => items.find((token) => token.tokenInfo.address === tokenAddress)?.tokenInfo)
+    .catch(() => null)
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/services/firebase/notifications.ts]: Addressed in bbe73f4277ad22d090b588b66f45f57e3d06d560.

File: src/services/firebase/notifications.ts (hunk 15 above)
Comment (iamacook) [reply src/services/firebase/notifications.ts]: Addressed in bbe73f4277ad22d090b588b66f45f57e3d06d560.

File: src/components/settings/PushNotifications/logic.ts (hunk 16 above)
Comment (iamacook) [reply src/components/settings/PushNotifications/logic.ts]: Addressed in bbe73f4277ad22d090b588b66f45f57e3d06d560.

**Comment by iamacook** [comment]
> I'm lacking clarity on which modules are used by a service worker and which are not. Can we either name them in a distinct way or maybe add a comment on top of each?

I've restructured the code, moving that which is relevant _only_ for the service worker to the `@/service-workers/firebase` folder, as well as rename functions accordingly in bbe73f4277ad22d090b588b66f45f57e3d06d560.

File: src/service-workers/firebase-messaging/notifications.ts [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+
+  return preferencesStore.preferences[type]
+}
+
+const GATEWAY_URL_PRODUCTION = process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'https://safe-client.safe.global'
+const GATEWAY_URL_STAGING = process.env.NEXT_PUBLIC_GATEWAY_URL_STAGING || 'https://safe-client.staging.5afe.dev'
+
+// localStorage cannot be accessed in service workers so we reference the flag from the environment
+const GATEWAY_URL = FIREBASE_IS_PRODUCTION ? GATEWAY_URL_PRODUCTION : GATEWAY_URL_STAGING
+setBaseUrl(GATEWAY_URL)
```
Comment (katspaugh) [inline src/service-workers/firebase-messaging/notifications.ts:47]: Could you move this to the top of the file?

File: src/service-workers/firebase-messaging/notifications.ts [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+        body: `Safe ${shortenAddress(
+          address,
+        )} on ${chainName} has a new confirmation request for transaction ${shortenAddress(safeTxHash)}.`,
+      }
+    },
+    [WebhookType.SAFE_CREATED]: () => {
+      // We do not preemptively subscribe to Safes before they are created
+      return null
+    },
+  }
```
Comment (katspaugh) [inline src/service-workers/firebase-messaging/notifications.ts:217]: This could be moved to a separate module as a static config.

File: src/service-workers/firebase-messaging/notifications.ts (hunk 18 above)
Comment (iamacook) [reply src/service-workers/firebase-messaging/notifications.ts]: Addressed in a115c14973ce9cf0f7b6af14d7ae633ef6af035d.

File: src/service-workers/firebase-messaging/notifications.ts (hunk 19 above)
Comment (iamacook) [reply src/service-workers/firebase-messaging/notifications.ts]: Moved in a115c14973ce9cf0f7b6af14d7ae633ef6af035d.

**Review by katspaugh — APPROVED** [review]

File: src/components/settings/PushNotifications/logic.ts [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+    logError(ErrorCodes._400, e)
+  }
+
+  const isGranted = permission === 'granted'
+
+  trackEvent(isGranted ? PUSH_NOTIFICATION_EVENTS.GRANT_PERMISSION : PUSH_NOTIFICATION_EVENTS.REJECT_PERMISSION)
+
+  if (!isGranted) {
+    alert('You must allow notifications to register your device.')
+  }
```
Comment (usame-algan) [inline src/components/settings/PushNotifications/logic.ts:38]: We show this alert on the global preferences but not when on a specific safe. Is there a reason for that?

File: src/components/settings/PushNotifications/logic.ts [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+  const BUILD_NUMBER = '0' // Required value, but does not exist on web
+  const BUNDLE = 'safe'
+
+  const [serviceWorkerRegistration] = await navigator.serviceWorker.getRegistrations()
+
+  // Get Firebase token
+  const app = initializeFirebaseApp()
+  const messaging = getMessaging(app)
+
+  const token = await getToken(messaging, {
```
Comment (usame-algan) [inline src/components/settings/PushNotifications/logic.ts:84]: The documentation of this function says: `If a notification permission isn't already granted, this method asks the user for permission.`

Isn't that what we also do manually in `requestNotificationPermission`?

File: src/hooks/useDecodeTx.ts [hunk 22]
```diff
@@ -14,7 +14,7 @@ const useDecodeTx = (tx?: SafeTransaction): AsyncResult<DecodedDataResponse> =>
   const nativeTransfer = isEmptyData && !isRejection ? getNativeTransferData(tx?.data) : undefined
 
   const [data = nativeTransfer, error, loading] = useAsync<DecodedDataResponse>(() => {
-    if (!encodedData || isEmptyData) return
+    if (!encodedData || isEmptyData || !tx.data.to) return
```
Comment (usame-algan) [inline src/hooks/useDecodeTx.ts:17]: The parameter should be optional now so I think we can remove this check.

File: package.json [hunk 23]
```diff
@@ -118,14 +120,17 @@
...
+    "patch-package": "^8.0.0",
     "pre-commit": "^1.2.2",
     "prettier": "^2.7.0",
     "ts-node": "^10.8.2",
     "ts-prune": "^0.10.3",
     "typechain": "^8.0.0",
     "typescript": "4.9.4",
-    "typescript-plugin-css-modules": "^4.2.2"
+    "typescript-plugin-css-modules": "^4.2.2",
+    "webpack": "^5.88.2"
```
[the remark this answers is not in the document]
Comment (iamacook) [reply package.json:134]: It is, but [`next-pwa` requires it](https://ducanh-next-pwa.vercel.app/docs/next-pwa/getting-started#install) to build service workers.

File: src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+    })
+  }, [dismissedBannerPerChain, safe.chainId, setDismissedBannerPerChain])
+
+  // Click outside to dismiss banner
+  useEffect(() => {
+    if (!shouldShowBanner) {
+      return
+    }
+
+    trackEvent(PUSH_NOTIFICATION_EVENTS.DISPLAY_BANNER)
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx:56]: Iirc, when I went over the required events with @kirkkonen, this was added to ease funnel creation.

It cannot be moved to `dismissBanner` we show the banner immediately.

File: src/components/settings/PushNotifications/index.tsx [hunk 25]
```diff
@@ -0,0 +1,263 @@
...
+
+                <FormControlLabel
+                  control={
+                    <Checkbox
+                      checked={preferences[WebhookType.NEW_CONFIRMATION]}
+                      onChange={(_, checked) => {
+                        setPreferences({
+                          ...preferences,
+                          [WebhookType.NEW_CONFIRMATION]: checked,
+                        })
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/index.tsx:206]: Yes. As `indexedDB` is promised base, we're only changing the state after it has resolved. We could change this to be optimistic but I think having the database as the SSOT makes most sense. Wdyt?

File: src/components/settings/PushNotifications/hooks/useNotificationPreferences.ts [hunk 26]
```diff
@@ -0,0 +1,217 @@
...
+
+  // Delete all preferences store entries
+  const clearPreferences = () => {
+    if (!preferencesStore) {
+      return
+    }
+
+    clearIndexedDb(preferencesStore)
+      .then(hydratePreferences)
+      .catch(() => null)
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/hooks/useNotificationPreferences.ts:205]: We only need to catch errors, the returned value isn't needed. However, as we have the `no-empty` TSLint rule on, an "empty" function won't work. Would you prefer I change them to `try / catch`?

File: src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+
+  const { registerNotifications } = useNotificationRegistrations()
+
+  const dismissBanner = useCallback(() => {
+    trackEvent(PUSH_NOTIFICATION_EVENTS.DISMISS_BANNER)
+
+    setDismissedBannerPerChain({
+      ...dismissedBannerPerChain,
+      [safe.chainId]: true,
+    })
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx]: Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

File: src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+  }, [dismissedBannerPerChain, safe.chainId, setDismissedBannerPerChain])
+
+  // Click outside to dismiss banner
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx:50]: Good point. We're using the same style as the batch banner atm. @liliiaorlenko, what are your thoughts?

File: src/components/settings/PushNotifications/index.tsx [hunk 29]
```diff
@@ -0,0 +1,263 @@
...
+    useNotificationRegistrations()
+
+  const preferences = getPreferences(safe.chainId, safe.address.value)
+
+  const setPreferences = (newPreferences: NonNullable<ReturnType<typeof getPreferences>>) => {
+    updatePreferences(safe.chainId, safe.address.value, newPreferences)
+  }
+
+  const isMac = typeof navigator !== 'undefined' && navigator.userAgent.includes('Mac')
+  const shouldShowMacHelper = isMac || IS_DEV
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/index.tsx]: I've created a new `useIsMac` hook in line with NextJS' usage of the `window` object in 12116b71f0067a697e64c861fc682219a9014108.

File: src/components/settings/PushNotifications/index.tsx [hunk 30]
```diff
@@ -0,0 +1,263 @@
...
+                            trackEvent({ ...PUSH_NOTIFICATION_EVENTS.TOGGLE_CONFIRMATION_REQUEST, label: checked })
+
+                            dispatch(
+                              showNotification({
+                                message:
+                                  'You will now receive notifications about confirmation requests for this Safe Account in your browser.',
+                                variant: 'success',
+                                groupKey: 'notifications',
+                              }),
+                            )
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/index.tsx]: Fixed in 12116b71f0067a697e64c861fc682219a9014108.

File: src/components/settings/PushNotifications/hooks/useNotificationPreferences.ts [hunk 31]
```diff
@@ -0,0 +1,217 @@
...
+      return safeAddresses.map((safeAddress) => getPushNotificationPrefsKey(chainId, safeAddress))
+    })
+
+    deleteManyFromIndexedDb(keysToDelete, preferencesStore)
+      .then(hydratePreferences)
+      .catch(() => null)
+  }
+
+  // Delete all preferences store entries
+  const clearPreferences = () => {
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/hooks/useNotificationPreferences.ts]: Renamed in 12116b71f0067a697e64c861fc682219a9014108.

File: src/components/settings/PushNotifications/PushNotificationsBanner/styles.module.css [hunk 32]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+.chip {
+  border-radius: 4px;
+  background-color: var(--color-secondary-main);
+  font-weight: 400;
+  font-size: 12px;
+  width: var(--space-5);
+  height: 24px;
+  z-index: 9999999;
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/PushNotificationsBanner/styles.module.css]: Indeed. Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

File: src/components/settings/PushNotifications/logic.ts (hunk 21 above)
Comment (iamacook) [reply src/components/settings/PushNotifications/logic.ts:74]: Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

File: src/components/settings/PushNotifications/logic.ts (hunk 20 above)
Comment (iamacook) [reply src/components/settings/PushNotifications/logic.ts]: I've removed this. The function is now being used as an early return and I've added a comment explaining it in 12116b71f0067a697e64c861fc682219a9014108.

File: src/components/settings/PushNotifications/logic.ts (hunk 21 above)
Comment (iamacook) [reply src/components/settings/PushNotifications/logic.ts:74]: See above comment.

File: src/hooks/useDecodeTx.ts (hunk 22 above)
Comment (iamacook) [reply src/hooks/useDecodeTx.ts]: Adjusted in 12116b71f0067a697e64c861fc682219a9014108.

File: src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx (hunk 24 above)
[the remark this answers is not in the document]
Comment (usame-algan) [reply src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx:56]: Maybe to track how many users see the banner but don't dismiss it? But since the banner is closed when clicking anywhere I don't think that would be a useful metric.

File: src/components/settings/PushNotifications/index.tsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (usame-algan) [reply src/components/settings/PushNotifications/index.tsx:206]: Agreed, lets not make it optimistic but maybe we can provide some loading indicator to the user.

File: src/components/settings/PushNotifications/hooks/useNotificationPreferences.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (usame-algan) [reply src/components/settings/PushNotifications/hooks/useNotificationPreferences.ts:205]: I think the current style is elegant but I am not sure if we should log anything in case those indexed db calls throw. I will do some research on the topic.

**Review by usame-algan — APPROVED** [review]
Works great and thanks for implementing the changes 👍

Will discuss the notification dismissal tomorrow but it is non-blocking imo. LGTM 🚀

File: src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx (hunk 28 above)
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx]: As per discussion, I've added a close button to the banner which also negates the above comment regarding the tracking.

image

File: src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx (hunk 24 above)
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/settings/PushNotifications/PushNotificationsBanner/index.tsx:56]: As per the addition of a close button, this is now not relevant.

**Comment by liliya-soroka** [comment]
Feedback: A notification banner is displayed for the safe with enabled notifications through Global settings. It's confusing, as I just enabled and signed notifications enabling for this safe 
image
Steps:
1. make sure that you have a few safes in the app on different networks
2. run app 
3. Select Enable Notifications using Global settings
4. enable notifications for all your safes 
5. come back to the safes list and switch to the safe on another network
Current result: Notifications banner is displayed

**Comment by francovenica** [comment]
Looks good.
The ticket  was checked by several people, in windows and mac for the different type of notifications. 
All the issues reported in the QA document were fixed or discussed: https://www.notion.so/safe-global/QA-Feedback-for-notifications-epic-2391-557125a8a91240a59903655e0cfc3bd3

A few notes for future iterations that we might want to revisit:
You get a notification for every tx executed in a "bulk transaction execution".
The "Confirm request" only works for safes with at least 3 owners and a policy of at least 2. I assume this is cuz for a threshold of 1 you don't need to confirm anymore, and if the safe has 2 owners, the tx 2nd confirmation needed is a "sign and execute" which I believe is different from a "just confirm" type of signing.
